core/src/main/java/com/opensymphony/xwork2/conversion/impl/DefaultConversionAnnotationProcessor.java

```diff
@@ -80,18 +81,22 @@ else if (tc.rule() != ConversionRule.ELEMENT || tc.rule() == ConversionRule.KEY
                 else if (tc.rule() == ConversionRule.KEY) {
                     Class<?> converterClass;
                     if (StringUtils.isNoneEmpty(tc.converter())) {
-                        converterClass = Thread.currentThread().getContextClassLoader().loadClass(tc.converter());
-                        //check if the converter is a type converter if it is one
-                        //then just put it in the map as is. Otherwise
-                        //put a value in for the type converter of the class
+                        converterClass = ClassLoaderUtil.loadClass(tc.converter(), this.getClass());
                     } else {
                         converterClass = tc.converterClass();
                     }
 
                     LOG.debug("Converter class: [{}]", converterClass);
 
+                    //check if the converter is a type converter if it is one
+                    //then just put it in the map as is. Otherwise
+                    //put a value in for the type converter of the class
                     if (converterClass.isAssignableFrom(TypeConverter.class)) {
-                        mapping.put(key, converterCreator.createTypeConverter(tc.converter()));
+                        if (StringUtils.isNoneEmpty(tc.converter())) {
+                            mapping.put(key, converterCreator.createTypeConverter(tc.converter()));
+                        } else {
+                            mapping.put(key, converterCreator.createTypeConverter(tc.converterClass()));
+                        }
```

Comment:
I afraid if user e.g. has a spring bean with name "java.lang.String" but with another type than string. In such cases, `converterCreator.createTypeConverter("java.lang.String")` and `converterCreator.createTypeConverter(java.lang.String.class)` will have different results :)

Reply:
Not sure if I understand your example with `String.class` - in this place there be only `TypeConverter.class`, not a string. The `.converter()` value was used above to load a class (line 84), on line 94 we check if the class is `TypeConverter.class` and in such case we create the converter. So in both case it must be an instance of `TypeConverter` class.

Reply:
Thanks! You're right. Please replace `java.lang.String` with `com.opensymphony.xwork2.conversion.TypeConverter` in my above comment. This is an odd thing :) but please consider a spring bean with name `"com.opensymphony.xwork2.conversion.TypeConverter"` and that bean type is `me.yasser.MyTypeConverter`. In such case,  `converterCreator.createTypeConverter("com.opensymphony.xwork2.conversion.TypeConverter")` and `converterCreator.createTypeConverter(com.opensymphony.xwork2.conversion.TypeConverter.class)` will have different results.

At bottom, I think this was not working already before these and previous changes. Because [DefaultObjectTypeDeterminer.getKeyClass](https://github.com/apache/struts/blob/6e96f11debc4fa52c65a12b28fea82b514b96abd/core/src/main/java/com/opensymphony/xwork2/conversion/impl/DefaultObjectTypeDeterminer.java#L96) always casts it to `Class` and will fail at this case when it's `TypeConverter`.

Reply:
>At bottom, I think this was not working already before these and previous changes. Because DefaultObjectTypeDeterminer.getKeyClass always casts it to Class and will fail at this case when it's TypeConverter.

FYI: I knew this piece of code was not working already before these and previous changes. But I just kept it unchanged as maybe I don't know something.

Reply:
Hm... still not sure if I follow (not a Spring fan) but it's safer to keep it as is if you still have some objections and corner cases.